/*
banks.proto
CSE 531 - CCC Read Your Writes Project
tfilewic
2025-11-27

Defines messages and services for the distributed banking system.
*/

syntax = "proto3";

//request to query the current balance
message BalanceRequest {
    repeated int32 writeset = 1;
}

//response to the balance request
message BalanceResponse {
    int32 balance = 1;
}

//customer request to make a deposit or withdrawal transaction
message TransactionRequest {
    int32 amount = 1;
}

//branch request to propagate a deposit or withdrawal transaction
message PropagationRequest {
    int32 amount = 1;
    int32 write_id = 2;
}

//response to the deposit or withdrawal transaction request
message TransactionResponse {
    int32 write_id = 1;    //0 for failure
}

//service defining all RPC interfaces for customers and branches
service RPC {
    rpc Query (BalanceRequest) returns (BalanceResponse);
    rpc Deposit (TransactionRequest) returns (TransactionResponse);
    rpc Withdraw (TransactionRequest) returns (TransactionResponse);
    rpc Propagate_Deposit (PropagationRequest) returns (TransactionResponse);
    rpc Propagate_Withdraw (PropagationRequest) returns (TransactionResponse);
}